### PR TITLE
dhall-nix: Fix field-based union access

### LINKED ISSUE #906: dhall-to-nix generate incorrect Nix for sum types
```
$ /nix/store/hi06lr4hfq52krhkr99vdfxda836mbzz-dhall-nix-1.1.6/bin/dhall-to-nix <<< '<Foo>.Foo'
{}.Foo

$ nix-instantiate --eval-only -E $( /nix/store/hi06lr4hfq52krhkr99vdfxda836mbzz-dhall-nix-1.1.6/bin/dhall-to-nix <<< '<Foo>.Foo' )
error: attribute 'Foo' missing, at (string):1:1
```

`/nix/store/hi06lr4hfq52krhkr99vdfxda836mbzz-dhall-nix-1.1.6` is the latest `linux-dhall-nix` build: https://hydra.dhall-lang.org/build/10712#tabs-details

Note this also happens with sum types with arguments:

```
$ /nix/store/hi06lr4hfq52krhkr99vdfxda836mbzz-dhall-nix-1.1.6/bin/dhall-to-nix <<< '<Foo : Bool >.Foo True'
{}.Foo true
```

### PR DESCRIPTION
**Pull request by ocharles**
Fixes #906.

I copied the approach used in `dhall-json`.

### CODE REVIEW

File: dhall-nix/src/Dhall/Nix.hs [hunk 1]
```diff
@@ -488,6 +495,12 @@ dhallToNix e = loop (Dhall.Core.normalize e)
         a' <- loop a
         b' <- loop b
         return (Fix (NBinary NUpdate a' b'))
+    loop (Field (Union kts) k) = do
+        let e0 = do
+                k' <- Dhall.Map.keys kts
+                return (k', Nothing)
+        let e2 = Fix (NSym k)
+        return (Fix (NAbs (ParamSet e0 False Nothing) e2))
```
Comment (Gabriel439) [inline dhall-nix/src/Dhall/Nix.hs]: If I understand correctly, this compiles `< Foo >.Foo` to `{ Foo }: Foo`, but I believe the result should instead be: `x: { Foo }: Foo x`

Comment (ocharles) [reply dhall-nix/src/Dhall/Nix.hs]: `<Foo>.Foo` compiling to `{Foo}: Foo` seems correct to me - what is that `x` about?

Comment (Gabriel439) [reply dhall-nix/src/Dhall/Nix.hs]: Oh, you're right.  Never mind, my thoughts were confused

Comment (Gabriel439) [reply dhall-nix/src/Dhall/Nix.hs]: Now I remember what led to my confusion.  The issue here is that you need to check if the alternative is empty when compiling the union to Nix.

For example:

```nix
dhallToNix "< Foo >.Foo" = { Foo }: Foo

dhallToNix "< Foo : {} >.Foo {=}" = { Foo }: Foo {}

dhallToNix "< Foo : {} >.Foo" = x: { Foo }: Foo x
```

Carefully note the difference between the 1st and 3rd examples.  If the type of the alternative is present then you need to turn it into an anonymous function.

Comment (ocharles) [reply dhall-nix/src/Dhall/Nix.hs]: Ah, so it's a case of `< .. >.x` being partially applied when the alternative does take data

Comment (ocharles) [reply dhall-nix/src/Dhall/Nix.hs]: Ok, I believe my latest commit resolves this.

**Comment by ocharles** [comment]
Here's the latest behaviour:

```
[nix-shell:~/work/dhall-haskell/dhall-nix]$ ./dist/build/dhall-to-nix/dhall-to-nix <<< '<Foo : {} | Bar>.Foo'
x:
  { Bar, Foo }:
    Foo x

[nix-shell:~/work/dhall-haskell/dhall-nix]$ ./dist/build/dhall-to-nix/dhall-to-nix <<< '<Foo : {} | Bar>.Foo {=}'
{ Bar, Foo }:
  Foo {}

[nix-shell:~/work/dhall-haskell/dhall-nix]$ ./dist/build/dhall-to-nix/dhall-to-nix <<< '<Foo : {} | Bar>.Bar'
{ Bar, Foo }:
  Bar
```

**Comment by Gabriel439** [comment]
@ocharles: I think you need to import `join`: https://hydra.dhall-lang.org/build/10741/nixlog/2/tail

**Comment by ocharles** [comment]
Ah no, I just need to push another commit

**Pull request merged**